File: 206-reverse-linked-list/reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        ListNode next = null;

        while (curr != null) {
            // store og next pointer
            next = curr.next;
            
            // reverse the link
            curr.next = prev;

            // update pointers
            prev = curr;
            curr = next;
        }

        // tail of old list becomes head of new list
        return prev;
    }
}

File: 206-reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        
        prev = None
        curr = head

        while curr:
            # store pointers before changing
            temp_next = curr.next

            # change references
            curr.next = prev

            # move pointers
            prev = curr
            curr = temp_next
        
        return prev

